miden-tx/build.rs

```diff
@@ -0,0 +1,173 @@
+use std::{collections::BTreeMap, env, fmt::Write, path::Path};
+
+use assembly::{
+    diagnostics::{IntoDiagnostic, Result},
+    Report,
+};
+use regex::Regex;
+use walkdir::WalkDir;
+
+const ASM_DIR: &str = "../miden-lib/asm";
+const VM_ERRORS_FILE: &str = "src/errors/vm_errors.rs";
+
+fn main() -> Result<()> {
+    // re-build when the MASM code changes
+    println!("cargo:rerun-if-changed={ASM_DIR}");
+    println!("cargo:rerun-if-changed={VM_ERRORS_FILE}");
+
+    // Copies the MASM code to the build directory
+    let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
+    let asm_dir = Path::new(&crate_dir).join(ASM_DIR);
+
+    // Generate kernel error constants.
+    generate_kernel_error_constants(&asm_dir)?;
+
+    Ok(())
+}
+
+// KERNEL ERROR CONSTANTS
+// ================================================================================================
+
+/// Reads all MASM files from the `kernel_source_dir` and extracts its error constants and their
+/// associated comment as the error message and generates a Rust file from them. For example:
+///
+/// ```text
+/// # New account must have an empty vault
+/// const.ERR_PROLOGUE_NEW_ACCOUNT_VAULT_MUST_BE_EMPTY=0x0002000F
+/// ```
+///
+/// would generate a Rust file with the following content:
+///
+/// ```rust
+/// pub const ERR_PROLOGUE_NEW_ACCOUNT_VAULT_MUST_BE_EMPTY: u32 = 0x0002000f;
+/// ```
+///
+/// and add an entry in the constant -> error mapping array:
+///
+/// ```rust
+/// (ERR_PROLOGUE_NEW_ACCOUNT_VAULT_MUST_BE_EMPTY, "New account must have an empty vault"),
+/// ```
+///
+/// The caveats are that only the comment line directly above the constant is considered an error
+/// message. This could be extended if needed, but for now all errors can be described in one line.
+///
+/// We also ensure that a constant is not defined twice, except if their error code is the same.
+/// This can happen across multiple files.
+fn generate_kernel_error_constants(kernel_source_dir: &Path) -> Result<()> {
```

Comment:
It would be awesome to check if somewhere we use the same error code for two different errors (similar how we check right now that the that two different error codes are not used for the same error).

Reply:
Good idea. We were already using two codes with separate names.

Reply:
What is still unclear to me is how we actually come up with the error codes.
One idea we could also add in the build script is to enforce that errors of a certain category are within a certain range, if that was the initial idea. `ERR_ACCOUNT_*` might be within `0x00020040 - 0x00020050`, `ERR_FUNGIBLE_ASSET_*` within `0x00020050 - 0x00020060` and so on (ranges chosen arbitrarily).

This would ultimately enable range checks on the error code to determine the error category which would be nice. I'm not sure what other aspects have flown into the error design so far though - if there's an issue/discussion somewhere I would be happy to catch up.

Reply:
Currently, the only pattern is that errors emitted by the transaction kernel should have the most significant 16 bits set to 2. But I do think it is a good idea to segment them further into categories. Let's create an issue for this (or update an existing issue if there is something already related to this).